File: Python3/215_Kth_Largest_Element_in_an_Array-quickSelect.py
class Solution:
    def findKthLargest(self, nums: List[int], k: int) -> int:
        
        def quickSelect(left, right, goal_pos):
            pivot = nums[right]
            j = left
            for i in range(left, right):
                if nums[i] > pivot:
                    nums[i], nums[j] = nums[j], nums[i]
                    j += 1
            nums[right], nums[j] = nums[j], nums[right]
            if j == goal_pos:
                return pivot
            elif j < goal_pos:
                return quickSelect(j+1, right, goal_pos) # go right
            
            return quickSelect(left, j-1, goal_pos) # go left
        
        return quickSelect(0, len(nums)-1, k-1)

File: C++/215_Kth_Largest_Element_in_an_Array-quickSelect.cpp
class Solution {
public:
    void swap(vector<int>& nums, int i, int j) {
        int t = nums[i];
        nums[i] = nums[j];
        nums[j] = t;
    }
    int partition(vector<int>& nums, int l, int r) {
        int pivot = nums[l];
        int p = l;
        while (l < r) {
            if (nums[r] < pivot) {
                r--;
                continue;
            }
            if (nums[l] >= pivot) {
                l++;
                continue;
            }
            swap(nums, l, r);
        }
        swap(nums, l, p);
        return l;
    }
    int findKthLargest(vector<int>& nums, int k) {
        int l = 0, r = nums.size() - 1;
        while (true) {
            int m = partition(nums, l, r);
            if (m == k - 1) return nums[m];
            else if (m > k - 1) r = m - 1;
            else l = m + 1;
        }
        return -1;
    }
};